Phase 4: Update  Documentation (#155)
Ekumen-OS/lambkin: src/lambkin/README.md

Reviewer: ```suggestion
@lambkin.option("--clock-rate", default=1)
```

Author: Solved in 105a3e3

Reviewer: ```suggestion
    with lambkin.process.background(ctx.shell.my_recorder, "-o", "output"):
```

Author: Solved in  105a3e3

Reviewer: I am missing a comment explaining or emphasizing the shell usage difference in background vs used standalone (parenthesis).

Author: Solved in a409d7d

Reviewer: Missing the metadata file here. Worth also adding log file examples.

Author: Solved in e26f77b

Abstracts shell command dispatch. Exposes the host environment's executables as Python attributes — accessing `shell.my_tool` returns a callable that runs `my_tool` with the given arguments, letting benchmark scripts invoke external processes without hardcoding paths or constructing subprocess calls manually. Accessible through the context.

**Background Process**

Runs a process in the background while the benchmark continues executing. Takes a shell command (without calling it) and manages its full lifecycle — start, monitor, and clean up — as a context manager. When the context exits, it terminates the process and all its descendants; any process spawned outside a `background()` block is not covered.

> [!WARNING]

> > ctx.shell builds commands lazily through attribute chaining — each attribute access appends a word to the command. Pass the proxy without calling it to `lambkin.process.background()`. Calling it with () executes it immediately as a foreground process before `background()` can manage it.

Reviewer: Sorry but this does not very read very clearly to me.

```suggestion
> [!WARNING]
> Pass the command proxy to background() without calling it — ctx.shell.my_tool, not ctx.shell.my_tool(). Calling it with () runs the process immediately as a foreground blocking call and background() will raise an error.
```

With this comment probably the example below is not needed.

Author: Solved in 819b039

```python
# Correct — proxy passed without calling it
with lambkin.process.background(
    ctx.shell.ros2.launch, "beluga_ros2", "beluga.launch.py"
):
    ...

# Wrong — calling it with () executes it immediately as a foreground process
with lambkin.process.background(

>     ctx.shell.ros2.launch(), "beluga_ros2", "beluga.launch.py"

Reviewer: ```suggestion
    ctx.shell.ros2.launch("beluga_ros2", "beluga.launch.py")
```

Author: This section had been deleted

):
    ...
```

**Process Isolation with cgroups v2**

> LAMBKIN places each iteration in its own cgroup, so every process spawned during that iteration — whether inside a `background()` block or not — is tracked and contained. This serves two purposes:

Reviewer: I don't think this is how you want this to look.

<img width="1081" height="277" alt="Image" src="https://github.com/user-attachments/assets/170f63f0-0b40-4415-a6d8-6e3a39a5aaac" />

Anyway, I think this diserves a section of its own, as it is not a core concept. Maybe a section named "Cgroups Process Management".

Author: Solved in 2a1243a

> * Isolation — processes from one iteration cannot bleed into another. If iteration N is still cleaning up when iteration N+1 starts, there is no interference.

Reviewer: This is misleading. The core idea here is that the actual guarantee is narrower: processes from a previous iteration won't survive into the next one because the cgroup is torn down on Context.__exit__. But that is already explained above.

Author: Solved in 09ede79

The cgroup hierarchy for a run looks like this:
```
app.slice/                               ← user's systemd app slice
└── lambkin-my_benchmark-a1b2c3d4/       ← one per CLI invocation
    └── iter-var_1-iter_1-e5f6a7b8/      ← one per (variant, iteration) pair
        ├── my_algorithm-a9b0c1d2/       ← ros2 launch process
        └── my_recorder-e3f4a5b6/        ← ros2 bag record process
```
When running on the host, a user systemd app slice (app.slice) is always available. In containerized environments no app slice may exist — in that case, LAMBKIN falls back to the nearest delegated cgroup it can find. For example, under Podman with `--systemd=always`:
```
user.slice/user-1000.slice/user@1000.service/  ← delegated cgroup root
└── lambkin-my_benchmark-a1b2c3d4/
    └── iter-var_1-iter_1-e5f6a7b8/
        ├── my_algorithm-a9b0c1d2/
        └── my_recorder-e3f4a5b6/
```

Cleanup on exit. When a background() context exits normally, LAMBKIN sends `SIGTERM` to all processes in the cgroup, waits for a grace period, then sends `SIGKILL` to any survivors. On Ctrl-C, the CLI writes 1 to `cgroup.kill`, which the kernel propagates instantly to the entire tree.

> [!NOTE]

> > The cgroup design provides process lifetime containment, not communication isolation. If iterations were to run in parallel, processes from different iterations could still communicate with each other.

Reviewer: Feels more like a warning

Author: Solved in fe356c4

## CLI

LAMBKIN exposes a lambkin command that runs your benchmark script inside a transient systemd cgroup scope, ensuring all child processes are tracked and cleaned up automatically.

```bash
Usage: lambkin [OPTIONS] SCRIPT [SDK_OPTIONS] [CUSTOM_OPTIONS]

Options:
  --help          Show this message and exit.

SDK Options (always available):
  --dry-run       Run the benchmark in dry-run mode: commands are logged but
                  not executed.
  --show-options  List all SDK and custom options available for this benchmark
                  script and exit.

Custom Options (script-defined):
  Options registered in your benchmark script via @lambkin.option.
  Run 'lambkin SCRIPT --show-options' to list them.
```

> ## Logging

Reviewer: I am missing the log level control over the sdk logs. Maybe you can separate here into 2 subsections for clarity. One that addressed SDK logging another that addresses Process logging.

Author: Solved in 060aa7e

Abstracts shell command dispatch. Exposes the host environment's executables as Python attributes — accessing `shell.my_tool` returns a callable that runs `my_tool` with the given arguments, letting benchmark scripts invoke external processes without hardcoding paths or constructing subprocess calls manually. Accessible through the context.

**Background Process**

Runs a process in the background while the benchmark continues executing. Takes a shell command (without calling it) and manages its full lifecycle — start, monitor, and clean up — as a context manager. When the context exits, it terminates the process and all its descendants; any process spawned outside a `background()` block is not covered.

> [!WARNING]

> > Pass the command proxy to background() without calling it — ctx.shell.my_tool, not ctx.shell.my_tool(). Calling it with () runs the process immediately as a foreground blocking call and background() will raise an error.

Reviewer: ```suggestion
> Pass the command proxy to background() without calling it — `ctx.shell.my_tool`, not `ctx.shell.my_tool()`. Calling it with () runs the process immediately as a foreground blocking call and background() will raise an error.
```

Author: Solved in 187391e

## Process Management with cgroups v2

> LAMBKIN places each iteration in its own cgroup, so every process spawned during that run — whether inside a background() block or not — is tracked and torn down unconditionally when the iteration ends. This means no leftover processes survive into the next iteration, and no process can escape by calling setsid or setpgid — the kernel enforces containment regardless.

Reviewer: ```suggestion
LAMBKIN places each iteration in its own cgroup, so every process spawned during that run — whether inside a background() block or not — is tracked and torn down unconditionally when the iteration ends. This means no leftover processes survive into the next iteration, and no process can escape by calling `setsid` or `setpgid` — the kernel enforces containment regardless.
```

Author: Solved in 187391e

The cgroup hierarchy for a run looks like this:
```
app.slice/                               ← user's systemd app slice
└── lambkin-my_benchmark-a1b2c3d4/       ← one per CLI invocation
    └── iter-var_1-iter_1-e5f6a7b8/      ← one per (variant, iteration) pair
        ├── my_algorithm-a9b0c1d2/       ← ros2 launch process
        └── my_recorder-e3f4a5b6/        ← ros2 bag record process
```
When running on the host, a user systemd app slice (app.slice) is always available. In containerized environments no app slice may exist — in that case, LAMBKIN falls back to the nearest delegated cgroup it can find. For example, under Podman with `--systemd=always`:
```
user.slice/user-1000.slice/user@1000.service/  ← delegated cgroup root
└── lambkin-my_benchmark-a1b2c3d4/
    └── iter-var_1-iter_1-e5f6a7b8/
        ├── my_algorithm-a9b0c1d2/
        └── my_recorder-e3f4a5b6/
```

> [!WARNING]
> The cgroup design provides process lifetime containment, not network isolation. If iterations were to run in parallel, processes from different iterations could still communicate with each other.

> ## Process Cleanup

Reviewer: This is part of the process management section
```suggestion
### Process Cleanup
```

Reviewer: Missing this  @teresa-ortega

Author: Solvd in 45265bc

> When a background() context exits normally, LAMBKIN sends `SIGTERM` to all processes in the cgroup, waits for a grace period, then sends `SIGKILL` to any survivors. On Ctrl-C, the CLI writes 1 to `cgroup.kill`, which the kernel propagates instantly to the entire tree.

Reviewer: Here we are not interested on how background exits but how the entire benchmark is cleaned up per run. Rephrase so we focus on that and not the background.

Author: Solved in 187391e

**Background Process**

Runs a process in the background while the benchmark continues executing. Takes a shell command (without calling it) and manages its full lifecycle — start, monitor, and clean up — as a context manager. When the context exits, it terminates the process and all its descendants; any process spawned outside a `background()` block is not covered.

> [!WARNING]
> Pass the command proxy to background() without calling it — ctx.shell.my_tool, not ctx.shell.my_tool(). Calling it with () runs the process immediately as a foreground blocking call and background() will raise an error.

## Process Management with cgroups v2

LAMBKIN places each iteration in its own cgroup, so every process spawned during that iteration — whether inside a `background()` block or not — is tracked and contained. This serves two purposes:

* Isolation — each iteration runs in a dedicated cgroup, so processes from different iterations cannot interfere with each other.
* Full-run cleanup guarantee — the iteration cgroup covers all processes spawned during that run, including foreground calls. Even a benchmark that never uses background() gets this guarantee for free.

> Unlike process groups or sessions, a process cannot escape its cgroup by calling `setsid` or `setpgid` — the kernel enforces containment regardless.

Reviewer: I think this is tighter and less redundant. And we do not orbit around the concept of isolation which can feel tricky here.
```suggestion
LAMBKIN places each iteration in its own cgroup, so every process spawned during that run — whether inside a background() block or not — is tracked and torn down unconditionally when the iteration ends. This means no leftover processes survive into the next iteration, and no process can escape by calling setsid or setpgid — the kernel enforces containment regardless.
```

Author: Solved in 075f241

The cgroup hierarchy for a run looks like this:
```
app.slice/                               ← user's systemd app slice
└── lambkin-my_benchmark-a1b2c3d4/       ← one per CLI invocation
    └── iter-var_1-iter_1-e5f6a7b8/      ← one per (variant, iteration) pair
        ├── my_algorithm-a9b0c1d2/       ← ros2 launch process
        └── my_recorder-e3f4a5b6/        ← ros2 bag record process
```
When running on the host, a user systemd app slice (app.slice) is always available. In containerized environments no app slice may exist — in that case, LAMBKIN falls back to the nearest delegated cgroup it can find. For example, under Podman with `--systemd=always`:
```
user.slice/user-1000.slice/user@1000.service/  ← delegated cgroup root
└── lambkin-my_benchmark-a1b2c3d4/
    └── iter-var_1-iter_1-e5f6a7b8/
        ├── my_algorithm-a9b0c1d2/
        └── my_recorder-e3f4a5b6/
```

> ## Process Cleanup

Reviewer: Add a space before the process cleanup title.

Author: Solved in f672e1d

When a background() context exits normally, LAMBKIN sends `SIGTERM` to all processes in the cgroup, waits for a grace period, then sends `SIGKILL` to any survivors. On Ctrl-C, the CLI writes 1 to `cgroup.kill`, which the kernel propagates instantly to the entire tree.

> [!WARNING]

> > The cgroup design provides process lifetime containment, not network isolation. If iterations were to run in parallel, processes from different iterations could still communicate with each other.

Reviewer: Move this block before process cleanup.

Author: Solved in f672e1d

### SDK Logging

Controls the verbosity of LAMBKIN's own internal messages via the `--log-level` SDK option. Accepts any level supported by [Python's logging](https://docs.python.org/3/library/logging.html#logging-levels) module, case-insensitive:
```bash
lambkin my_benchmark.py --log-level debug
lambkin my_benchmark.py --log-level DEBUG  # equivalent
```

The LAMBKIN logger is fully isolated from the root logger — user scripts can configure their own logging without any interference.

### Process Logging

Controls where subprocess stdout and stderr are routed. Each process can be configured independently via log_output.
```bash
lambkin my_benchmark.py --log-output console
```

| Mode | Behavior |
|------|-----------|
| `"console"` | Routes subprocess stdout/stderr to the terminal |
| `"file"` | Writes subprocess output to a per-process log file under the iteration output directory |

Log files are named after the command and written to the iteration directory:
```
results/var_1/iter_1/
├── my_algorithm.stdout.log
├── my_algorithm.stderr.log
├── my_recorder.stdout.log
└── my_recorder.stderr.log
```

> Precedence (highest to lowest)

Reviewer: ```suggestion
Precedence (highest to lowest):
```

Author: Solved in a6a5965

Abstracts shell command dispatch. Exposes the host environment's executables as Python attributes — accessing `shell.my_tool` returns a callable that runs `my_tool` with the given arguments, letting benchmark scripts invoke external processes without hardcoding paths or constructing subprocess calls manually. Accessible through the context.

**Background Process**

Runs a process in the background while the benchmark continues executing. Takes a shell command (without calling it) and manages its full lifecycle — start, monitor, and clean up — as a context manager. When the context exits, it terminates the process and all its descendants; any process spawned outside a `background()` block is not covered.

> [!WARNING]
> Pass the command proxy to background() without calling it — ctx.shell.my_tool, not ctx.shell.my_tool(). Calling it with () runs the process immediately as a foreground blocking call and background() will raise an error.

> ## Process Isolation with cgroups v2

Reviewer: ```suggestion
## Process Management with cgroups v2
```

Author: Solved in ab66d53

LAMBKIN places each iteration in its own cgroup, so every process spawned during that iteration — whether inside a `background()` block or not — is tracked and contained. This serves two purposes:

* Isolation — each iteration runs in its own cgroup, which is torn down when the iteration context exits. Processes from one iteration cannot survive into the next.

> * Full-run cleanup guarantee — the iteration cgroup covers all processes spawned during the run, including foreground calls. Even a benchmark that never uses `background()` gets this guarantee for free.

Reviewer: The "Isolation" bullet is still slightly misleading — "each iteration runs in its own cgroup" is just restating the opening sentence, and "cannot survive into the next" is really a cleanup guarantee, not isolation. The two bullets are still blurring into each other.

Author: Solved in 7203bcd

Unlike process groups or sessions, a process cannot escape its cgroup by calling `setsid` or `setpgid` — the kernel enforces containment regardless.

The cgroup hierarchy for a run looks like this:
```
app.slice/                               ← user's systemd app slice
└── lambkin-my_benchmark-a1b2c3d4/       ← one per CLI invocation
    └── iter-var_1-iter_1-e5f6a7b8/      ← one per (variant, iteration) pair
        ├── my_algorithm-a9b0c1d2/       ← ros2 launch process
        └── my_recorder-e3f4a5b6/        ← ros2 bag record process
```
When running on the host, a user systemd app slice (app.slice) is always available. In containerized environments no app slice may exist — in that case, LAMBKIN falls back to the nearest delegated cgroup it can find. For example, under Podman with `--systemd=always`:
```
user.slice/user-1000.slice/user@1000.service/  ← delegated cgroup root
└── lambkin-my_benchmark-a1b2c3d4/
    └── iter-var_1-iter_1-e5f6a7b8/
        ├── my_algorithm-a9b0c1d2/
        └── my_recorder-e3f4a5b6/
```

> Cleanup on exit. When a background() context exits normally, LAMBKIN sends `SIGTERM` to all processes in the cgroup, waits for a grace period, then sends `SIGKILL` to any survivors. On Ctrl-C, the CLI writes 1 to `cgroup.kill`, which the kernel propagates instantly to the entire tree.

Reviewer: Regarding block
> Cleanup on exit. When a background() context exits normally, LAMBKIN sends `SIGTERM` to all processes in the cgroup, waits for a grace period, then sends `SIGKILL` to any survivors. On Ctrl-C, the CLI writes 1 to `cgroup.kill`, which the kernel propagates instantly to the entire tree.

I would promote to a heading of its own under 

```readme
### Process cleanup

My paragraph ...
```

And focus on the iteration cgroup teardown rather than the background.

Author: Solved in bf8384b

> [!WARNING]

> > The cgroup design provides process lifetime containment, not communication isolation. If iterations were to run in parallel, processes from different iterations could still communicate with each other.

Reviewer: ```suggestion
> The cgroup design provides process lifetime containment, not network isolation. If iterations were to run in parallel, processes from different iterations could still communicate with each other.
```

Author: Solved in 835020a

## CLI

LAMBKIN exposes a lambkin command that runs your benchmark script inside a transient systemd cgroup scope, ensuring all child processes are tracked and cleaned up automatically.

```bash
Usage: lambkin [OPTIONS] SCRIPT [SDK_OPTIONS] [CUSTOM_OPTIONS]

LAMBKIN is a benchmarking SDK for robotics applications. It runs your
benchmark script inside a dedicated cgroup v2 scope, ensuring all child
processes are tracked and cleaned up automatically.

Options:
  --help  Show this message and exit.

SDK Options (always available):
  --dry-run       Run the benchmark in dry-run mode: commands are logged but
                  not executed.
  --show-options  List all options registered via @lambkin.option.
  --log-output    Where to route process output: file or console.
  --log-level     Log level for lambkin SDK output.

Custom Options (script-defined):
  Options registered in your benchmark script via @lambkin.option.
  Run 'lambkin SCRIPT --show-options' to list them.
```

## Logging

LAMBKIN has two independent logging systems: one for its own internal messages and one for subprocess output.

**SDK Logging**

> Controls the verbosity of LAMBKIN's own internal messages. Set via the --log-level SDK option:

Reviewer: ```suggestion
### SDK Logging

Controls the verbosity of LAMBKIN's own internal messages. Set via the `--log-level` SDK option:
```

Author: Solved in d571c66

```bash
lambkin my_benchmark.py --log-level debug
```

| Level | Behavior |
|------|-----------|
| `"debug"` | All internal messages visible, including cgroup paths and command dispatch details |
| `"info"` | Standard operational messages visible (default) |

> | `"info"` | LAMBKIN output suppressed; only warnings and errors shown |

Reviewer: Info twice? Also, there are more than 3 levels: debug, info, warning, error, critical, ...

Author: Solved in ee780c6

> | `"info"` | LAMBKIN output suppressed; only warnings and errors shown |

Reviewer: I am not sure we need to explain the levels of a library that is not ours and is well documented. I would just mention that the logging log levels that can be passed via cli are the ones logging library supports and maybe add a link to their levels documentation. For example https://docs.python.org/3/library/logging.html#logging-levels

maybe mention if it is case sensitive or not. Like if you can pass --log-level DEBUG or --log-level debug

Author: Solved in ee780c6

The LAMBKIN logger is fully isolated from the root logger — user scripts can configure their own logging without any interference.

> **Process Logging**

Reviewer: ```suggestion
### Process Logging
```

Author: Solved in ee61fa3

Controls where subprocess stdout and stderr are routed. Each process can be configured independently via log_output.
```bash
lambkin my_benchmark.py --log-output console
```

| Mode | Behavior |
|------|-----------|
| `"console"` | Routes subprocess stdout/stderr to the terminal |
| `"file"` | Writes subprocess output to a per-process log file under the iteration output directory |

Log files are named after the command and written to the iteration directory:
```
results/var_1/iter_1/
├── my_algorithm.stdout.log
├── my_algorithm.stderr.log
├── my_recorder.stdout.log
└── my_recorder.stderr.log
```

> #### Precedence (highest to lowest)

Reviewer: I don't think this needs a section.
```suggestion
Precedence (highest to lowest):
```

Author: Solved in 2b2d956

Abstracts shell command dispatch. Exposes the host environment's executables as Python attributes — accessing `shell.my_tool` returns a callable that runs `my_tool` with the given arguments, letting benchmark scripts invoke external processes without hardcoding paths or constructing subprocess calls manually. Accessible through the context.

**Background Process**

> A context manager that wraps a Shell command and manages its full lifecycle — start, monitor, and clean up — ensuring no orphaned processes survive when the benchmark ends or is interrupted. Uses cgroups v2 to guarantee kernel-level cleanup of the entire process tree, including descendants that have detached via setsid or setpgid. Used via lambkin.process.background(...)

Reviewer: ```suggestion
A context manager that wraps a Shell command and manages its full lifecycle — start, monitor, and clean up — ensuring no orphaned processes survive when the benchmark ends or is interrupted. Uses cgroups v2 to guarantee kernel-level cleanup of the entire process tree, including descendants that have detached via setsid or setpgid. Used via `lambkin.process.background(...)`
```

Author: Solved in 105a3e3

**Process Isolation with cgroups v2**

> LAMBKIN uses cgroups v2 to track and clean up every process spawned during a benchmark run. Unlike process groups or sessions, a process cannot escape its cgroup by calling setsid or setpgid — the kernel enforces containment regardless of what the process does. This makes it the only reliable mechanism for cleaning up a full ROS 2 process tree.

Reviewer: I wouldn't mention a "full ROS 2 process tree". Remember LAMBKIN is ROS agnostic. a ROS 2 process tree is just an example but not the only one.

Author: Solved in 620ffde

> LAMBKIN uses cgroups v2 to track and clean up every process spawned during a benchmark run. Unlike process groups or sessions, a process cannot escape its cgroup by calling setsid or setpgid — the kernel enforces containment regardless of what the process does. This makes it the only reliable mechanism for cleaning up a full ROS 2 process tree.

Reviewer: Split these two sections more cleanly  (Background and Process Isolation with cgroups v2), as they currently overlap and leave gaps.

**BackgroundProcess** should lead with what it actually does for the user: it lets you run a process in background while the benchmark continues executing — that's its primary purpose and it's not stated. Then clarify the cleanup scope: it terminates the process and its descendants when the context exits, but only that subtree. A process spawned inside a background() block that itself forks children is covered; anything else in the iteration is not. The current wording 'ensures no orphaned processes survive' overpromises — if a process is launched outside a background() context manager, this class does nothing for it.

**Process Isolation with cgroups v2** should own two distinct ideas that are currently missing or blurred:
- Isolation: each iteration runs in its own cgroup, meaning processes from one iteration cannot bleed into another. If iteration N is still cleaning up when iteration N+1 starts, there is no interference. This justifies why we create a per-iteration cgroup rather than relying solely on BackgroundProcess.
- Full-run cleanup guarantee: the iteration cgroup covers every process spawned during that run, not just those wrapped in background(). A foreground call still lands in the iteration cgroup. This is the safety net that BackgroundProcess alone cannot provide, and it's the most important thing a user should understand about the cgroup design — even a benchmark that never calls background() gets this guarantee for free.

Reviewer: Worth highlighting:  the cgroup design gives you process lifetime containment, not communication isolation. So if we run iterations in parallel, there will be cross talk in between nodes.

Author: Solved in 52d4a25

Reviewer: I don't think this is is fully addressed. The first comment has not been taken into account and the second one is added as a note that is not properly visualized.

Author: Solved in e230fce

The cgroup hierarchy for a run looks like this:
```

> app.slice/                               ← user's systemd app slice

Reviewer: Mention that when no user app slice is found we attempt to find a delegated cgroup. And put an example. Running in the host guarantees a user systemd app slice being available but when running in containerized environments...

Author: Solved in 620ffde

                  not executed.
  --show-options  List all SDK and custom options available for this benchmark
                  script and exit.

Custom Options (script-defined):
  Options registered in your benchmark script via @lambkin.option.
  Run 'lambkin SCRIPT --show-options' to list them.
```

## Logging

LAMBKIN uses Python's standard logging module for its own informational messages. Subprocess output is handled separately through output redirection — each process can be configured independently with `log_output`.

**Output modes**

| Mode | Behavior |
|------|-----------|
| `"console"` | Routes subprocess stdout/stderr to the terminal |
| `"file"` | Writes subprocess output to a per-process log file under the iteration output directory |

Log files are named after the command and written to the iteration directory:
```
results/var_1/iter_1/
├── ros2_launch.stdout.log
├── ros2_launch.stderr.log
├── ros2_bag_record.stdout.log
└── ros2_bag_record.stderr.log
```

**Precedence** (highest to lowest)

1. **Per-call** — `_log_output` keyword at the call site
2. **Benchmark option** — `@lambkin.option("--log-output", default="file")`

> 3. **ShellProxy default** — `ShellProxy(log_output="file")`

Reviewer: - I think the precedence is not right. Isn't CLI the highest?
- log output is not registered via @lambkin.option as it is an SDK integrated option.
- log_output in per call has a leading slash: `_log_output` should be `log_output`

Author: Solved in 620ffde

```python
import lambkin

@lambkin.benchmark(
    variants=lambkin.common.named_product(
        param_a=["x", "y"],
        param_b=[1, 10, 100],
    ),
    num_iterations=5,
)
@lambkin.option("--clock-rate", default=1.0)
def my_benchmark(ctx):

>     with lambkin.process.background(ctx.shell.my_recorder, "-", "output.mcap"):

Reviewer: What happened with -o?
```suggestion
    with lambkin.process.background(ctx.shell.my_recorder, "-o", "output.mcap"):
```

Author: Solved in 347267c

        ):
            ctx.shell.my_player(ctx.inputs.dataset, "-r", ctx.options.clock_rate)

@my_benchmark.input
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / "my_dataset.mcap"

if __name__ == "__main__":
    my_benchmark()
```

> **Run it with the CLI:**

Reviewer: Be consistent with the style below:
```suggestion
Run it with the CLI:
```

Author: Solved in 5ddea14
